pkg/config/template.go

```diff
@@ -103,43 +108,53 @@ func (c *TemplateCMManager) CreateCMTemplates(templates CiTemplates) error {
 	return kutilerrors.NewAggregate(errors)
 }
 
-func (c *TemplateCMManager) CreateClusterProfiles(dir string, profiles []ClusterProfile) error {
-	var errs []error
-	for _, p := range profiles {
-		cm, err := genClusterProfileCM(dir, p)
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
-		c.logger.WithFields(logrus.Fields{"cluster-profile": cm.ObjectMeta.Name}).Info("creating rehearsal cluster profile ConfigMap")
-		if err := c.createCM(cm); err != nil {
-			errs = append(errs, err)
-		}
-	}
-	return kutilerrors.NewAggregate(errs)
+type osFileGetter struct {
+	root string
 }
 
-func genClusterProfileCM(dir string, profile ClusterProfile) (*v1.ConfigMap, error) {
-	ret := &v1.ConfigMap{
-		ObjectMeta: metav1.ObjectMeta{Name: profile.CMName()},
-		Data:       map[string]string{},
-	}
-	profilePath := filepath.Join(dir, profile.Name)
-	err := filepath.Walk(profilePath, func(path string, info os.FileInfo, err error) error {
-		if err != nil || info.IsDir() {
+func (g osFileGetter) GetFile(filename string) ([]byte, error) {
+	return ioutil.ReadFile(filepath.Join(g.root, filename))
+}
+
+func (c *TemplateCMManager) CreateClusterProfiles(dir string, cms map[string]prowplugins.ConfigMapSpec, profiles []ClusterProfile) error {
+	changes := []prowgithub.PullRequestChange{}
+	for _, profile := range profiles {
+		err := filepath.Walk(filepath.Join(dir, ClusterProfilesPath, profile.Name), func(path string, info os.FileInfo, err error) error {
+			if err != nil || info.IsDir() {
+				return err
+			}
+			if path, err = filepath.Rel(dir, path); err == nil {
+				changes = append(changes, prowgithub.PullRequestChange{
+					Filename: path,
+					Status:   string(prowgithub.PullRequestFileModified),
+				})
+			}
 			return err
-		}
-		b, err := ioutil.ReadFile(path)
+		})
 		if err != nil {
 			return err
 		}
-		ret.Data[filepath.Base(path)] = string(b)
-		return nil
-	})
-	if err != nil {
-		return nil, err
 	}
-	return ret, nil
+	var errs []error
+	for cm, data := range updateconfig.FilterChanges(cms, changes, c.logger) {
+		profile := strings.TrimPrefix(cm.Name, "cluster-profile-")
+		for _, p := range profiles {
+			if p.Name == profile {
+				profile = p.CMName()
+				break
+			}
+		}
+		err := c.createCM(&v1.ConfigMap{
```

Comment:
Update should do creation, why do we need this?

Reply:
To add the labels. This is something we could easily add to the upstream library in the future.

Reply:
Can we rely on upstream to create and just add labels ourselves after?

Reply:
That was the other option, but it would require a `Get` + `Update` here (`updateconfig.Update` doesn't return what it created), while `updateconfig` already does a `Get` first.

Reply:
For reference: https://github.com/kubernetes/test-infra/blob/50de43197c2fdf95570eb414e3510a0b562704f9/prow/plugins/updateconfig/updateconfig.go#L90

Reply:
Doing that also leaves a (very unlikely, but still…) window where we could create CMs that would never be removed if we failed before labeling them.